File: common-test/src/main/java/io/quartic/common/test/websocket/WebsocketServerRule.java
package io.quartic.common.test.websocket;

import org.glassfish.tyrus.server.TyrusServerContainer;
import org.glassfish.tyrus.spi.ServerContainer;
import org.junit.rules.ExternalResource;

import javax.websocket.DeploymentException;
import javax.websocket.Endpoint;
import javax.websocket.EndpointConfig;
import javax.websocket.Session;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static com.google.common.collect.Lists.newArrayList;
import static io.quartic.common.websocket.WebsocketUtilsKt.serverEndpointConfig;
import static java.util.Arrays.asList;
import static org.glassfish.tyrus.spi.ServerContainerFactory.createServerContainer;

public class WebsocketServerRule extends ExternalResource {
    private ServerContainer container;
    private int port;

    private final AtomicInteger numConnections = new AtomicInteger();
    private final List<String> messages = newArrayList();

    public class DummyEndpoint extends Endpoint {
        @Override
        public void onOpen(Session session, EndpointConfig config) {
            numConnections.incrementAndGet();
            messages.forEach(m -> {
                try {
                    session.getBasicRemote().sendText(m);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
        }
    }

    public String uri() {
        return "ws://localhost:" + port + "/ws";
    }

    @Override
    protected void before() throws Throwable {
        try {
            container = createServerContainer(null);
            container.addEndpoint(serverEndpointConfig("/ws", new DummyEndpoint()));
            container.start("", 0);
            port = ((TyrusServerContainer) container).getPort();
        } catch (IOException e) {
            throw new DeploymentException(e.getMessage(), e);
        }
    }

    @Override
    protected void after() {
        if (container != null) {
            container.stop();
            container = null;
            port = 0;
        }
    }

    public int numConnections() {
        return numConnections.get();
    }

    public void setMessages(String... messages) {
        setMessages(asList(messages));
    }

    public void setMessages(List<String> messages) {
        this.messages.clear();
        this.messages.addAll(messages);
    }
}

File: common-test/src/main/kotlin/io/quartic/common/test/websocket/WebsocketServerRule.kt
package io.quartic.common.test.websocket

import io.quartic.common.logging.logger
import io.quartic.common.websocket.serverEndpointConfig
import org.glassfish.tyrus.server.TyrusServerContainer
import org.glassfish.tyrus.spi.ServerContainerFactory.createServerContainer
import org.junit.rules.ExternalResource
import java.net.URI
import java.util.concurrent.atomic.AtomicInteger
import javax.websocket.*

// TODO - reimplement with coroutines so that we can await things
class WebsocketServerRule : ExternalResource() {
    private val LOG by logger()

    private var container: TyrusServerContainer? = null
    private val _numConnections = AtomicInteger()
    private val _numDisconnections = AtomicInteger()
    private val sessions = mutableListOf<Session>()
    var messages = emptyList<String>()
    val uri by lazy { URI("ws://localhost:${container!!.port}/ws") }
    val numConnections: Int get() = _numConnections.get()
    val numDisconnections: Int get() = _numDisconnections.get()
    val receivedMessages = mutableListOf<String>()

    override fun before() {
        container = createServerContainer(null) as TyrusServerContainer
        with (container!!) {
            addEndpoint(serverEndpointConfig("/ws", DummyEndpoint()))
            start("", 0)
        }
        LOG.info("Server listening at ${uri}")
    }

    override fun after() = stop()

    fun stop() {
        container?.stop()
        container = null
    }

    @Synchronized
    fun dropConnections() {
        sessions.forEach { it.close() }
    }

    @Synchronized
    private fun addConnection(session: Session) {
        sessions.add(session)
    }

    inner class DummyEndpoint : Endpoint() {
        override fun onOpen(session: Session, config: EndpointConfig) {
            LOG.info("onOpen")

            addConnection(session)
            session.addMessageHandler(MessageHandler.Whole<String> {
                LOG.info("Received message")
                receivedMessages.add(it)
            })
            _numConnections.incrementAndGet()
            messages.forEach { m -> session.basicRemote.sendText(m) }
        }

        override fun onClose(session: Session, closeReason: CloseReason) {
            LOG.info("onClose")
            _numDisconnections.incrementAndGet()
        }
    }
}
